runtime/download.go

```diff
@@ -0,0 +1,120 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/user"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/yourbase/yb/plumbing/log"
+)
+
+const defaultCacheDir = "/tmp/yourbase/cache"
+
+func localCacheDir() string {
+	if cacheDir, exists := os.LookupEnv("YB_CACHE_DIR"); exists {
+		return cacheDir
+	}
+	// Resolve current homeDir
+	if u, err := user.Current(); err == nil && u.HomeDir != "" {
+		return filepath.Join(u.HomeDir, ".cache", "yourbase")
+	}
+	// Try again
+	if homeDir, exists := os.LookupEnv("HOME"); exists {
+		return filepath.Join(homeDir, ".cache", "yourbase")
+	}
+	return defaultCacheDir
+}
+
+func cacheFilenameForURL(url string) (string, error) {
+	reg, err := regexp.Compile("[^a-zA-Z0-9.]+")
+	if err != nil {
+		return "", fmt.Errorf("Can't compile regex: %v", err)
+	}
+	fileName := reg.ReplaceAllString(url, "")
+	return fileName, nil
+}
+
+func downloadFileWithCache(ctx context.Context, url string) (string, error) {
```

Comment:
Optional: consider adding unit tests for this function. You can fake an HTTP server with [`httptest.Server`](https://pkg.go.dev/net/http/httptest?tab=doc#Server).